### PR TITLE
New Lint: Manual Flatten

### PR DESCRIPTION
**Pull request by nahuakang**
This is a draft PR for [Issue 6564](https://github.com/rust-lang/rust-clippy/issues/6564).

r? @camsteffen 

- \[x] Followed [lint naming conventions][lint_naming]
- \[x] Added passing UI tests (including committed `.stderr` file)
- \[x] `cargo test` passes locally
- \[x] Executed `cargo dev update_lints`
- \[x] Added lint documentation
- \[x] Run `cargo dev fmt`

---

*Please write a short comment explaining your change (or "none" for internal only changes)*
changelog: Add new lint [`manual_flatten`] to check for loops over a single `if let` expression with `Result` or `Option`.

### CODE REVIEW

**Comment by nahuakang** [comment]
@camsteffen Worked to this barebone version that checks for `for loops` and a single `if let` statement on `Option`/`Result`.

I have a few questions I'd like to ask for help on:

1. Is there a way I can check if the `arg` is an `Iterator` or implements `IntoIterator` so as to adjust the hint to cater to the `arg` type? Would it make sense to adjust the lint this way?
2. For using `span_lint_and_help`, is there a way to highlight the `if let` statement as well as the `arg`? 

Thanks a lot :pray:

**Comment by camsteffen** [comment]
Good start!

> r? @camsteffen

I'm not sure if I'm qualified as a reviewer around here 😄 but I'm glad to help you with this!

> Is there a way I can check if the `arg` is an `Iterator` or implements `IntoIterator` so as to adjust the hint to cater to the `arg` type? Would it make sense to adjust the lint this way?

You mean for `x` in `if let Some(y) = x`? Hmm I don't think it will help to check `x` in that way. Once we see `if let Some`, we already know that `Option` implements `IntoIterator`. If some unrecognized enum implements `IntoIterator`, we can't assume it is implemented in a way that `flatten()` would be equivalent.
 
[Here is a good example for what you are doing.](https://github.com/rust-lang/rust-clippy/blob/583715f583bd28ec759a5d0737b4ec400f0814cd/clippy_lints/src/needless_question_mark.rs#L163)

> For using `span_lint_and_help`, is there a way to highlight the `if let` statement as well as the `arg`?

Sure. That method takes two `Span`s. Just use the `Span` from the `Expr` with `ExprKind::Match` and then the `Span` from the `Expr` inside: `ExprKind::Match(Expr, ..)`. I assume the "help" would be "this implements `IntoIterator`" which is great!

Finally, I'd suggest naming the lint `manual_flatten`. The lint name typically should describe what is wrong with the code.

**Comment by camsteffen** [comment]
To clarify about the lint name, a for loop over `Option` is not inherently "bad". It is only "bad" if `None`s are just skipped over.

**Comment by flip1995** [comment]
> I'm not sure if I'm qualified as a reviewer around here

@camsteffen The question is if you want to get qualified as a reviewer here. :) 

I think this is a "good-first-review" PR. If you think this PR is ready, just ping me to take a quick look and then merge it :+1:

**Comment by nahuakang** [comment]
@camsteffen Hey Cam :) Thanks for the help. I did not explain myself well (and perhaps don't understand the language well enough either). There are two scenarios from what I can see:
```rust
// Scenario 1: probably a rare case
for n in vec![Some(1), Some(2), Some(3)].iter() {
    if let Some(n) = n {
        println!("{}", n);
    }
}

// Scenario 2
let x = vec![Some(1), Some(2), Some(3)];
for n in x {
    if let Some(n) = n {
        println!("{}", n);
    }
}
```
For the first one, if we recommend `flatten`, we'd only need to say:
> Consider using `flatten` on the Iterator: `vec![Some(1), Some(2), Some(3)].iter().flatten()`
For the second one, if we recommend `flatten`, we'd need to say:
> `x` implements `IntoIterator`. Consider using `flatten` on the Iterator: `x.iter().flatten()`.

So my first original question was how we could know if the type for the `arg` as in `for ... in arg` is an `Iterator` or it implements `IntoIterator`?

**Comment by camsteffen** [comment]
Ah right, that is a little tricky.

In `for x in iter`, `iter` implements `IntoIterator` as a rule. The question is whether `iter` is already an `Iterator` or if there is an implicit call to `into_iter()`. Implicit adjustments like this are called..."adjustments". You can get the type of an expression pre-adjustment using `cx.typeck_results().expr_ty(_)` (rather than `expr_ty_adjusted`). Check whether that type implements `Iterator` using `get_trait_def_id` and `implements_trait` from `utils`.

**Comment by nahuakang** [comment]
@camsteffen Check out [commit 005b0fd](https://github.com/rust-lang/rust-clippy/pull/6646/commits/005b0fd46cb523167c8f68d748f0f09864e9762d). The new lint already brought an internal improvement via dogfood :dog:

File: clippy_lints/src/loops.rs [hunk 1]
```diff
@@ -1953,6 +1986,54 @@ fn check_for_single_element_loop<'tcx>(
...
+                "Option"
+            } else {
+                "Result"
+            };
+            let msg = format!("for loop over `{}`s with a single `if let` expression.", pat_type);
+
+            // Prepare the help message
+            let arg_snippet = snippet(cx, arg.span, "..");
+            let arg_ty = cx.typeck_results().expr_ty(arg);
+            let is_iterator = get_trait_def_id(cx, &paths::ITERATOR).map_or(
```
Comment (camsteffen) [inline clippy_lints/src/loops.rs:2016]: Do `if let Some = get_trait_def_id()` above (bail out if `None`).

File: clippy_lints/src/loops.rs [hunk 2]
```diff
@@ -1953,6 +1986,54 @@ fn check_for_single_element_loop<'tcx>(
...
+        if pat_is_option || pat_is_result;
+
+        then {
+            // Prepare the error message
+            let pat_type = if pat_is_option {
+                "Option"
+            } else {
+                "Result"
+            };
+            let msg = format!("for loop over `{}`s with a single `if let` expression.", pat_type);
```
Comment (camsteffen) [inline clippy_lints/src/loops.rs:2011]: For the message and help message, let me suggest... (feel free to disagree 😃)

> Unnecessary `if let` since only the `Some` variant of the iterator element is used
> ...
> help: try `x.iter().flatten()`

**Comment by camsteffen** [comment]
Add a test for `if let None = ..` inside the loop (should not lint). I think you will have to adjust for this. See the link I posted [above](#issuecomment-768352270) for an example of detecting a `Some` constructor.

I found out that you can get the `Span` for the entire for-loop by using `arms[0].span` in `higher::for_loop`. If you feel so inclined, you can adjust that function to additionally return a `Span`.

File: clippy_lints/src/loops.rs (hunk 1 above)
Comment (nahuakang) [reply clippy_lints/src/loops.rs:2016]: Do you mean something like:
```rust
if let Some(id) = get_trait_def_id(cx, &paths::ITERATOR) {
    let is_iterator = implements_trait(cx, arg_ty, id, &[]);
    ...
    span_lint_and_help(....);            
} else {
    return;
}
```

Comment (camsteffen) [reply clippy_lints/src/loops.rs:2016]: Yes except put it in the `if_chain!` so that you don't need else, return.

**Comment by camsteffen** [comment]
Oh you need to suggest `into_iter()` instead of `iter()`.

File: clippy_lints/src/loops.rs [hunk 3]
```diff
@@ -1953,6 +1987,63 @@ fn check_for_single_element_loop<'tcx>(
...
+        // Ensure the `if let` statement is the only expression in the for-loop
+        if let ExprKind::Block(ref block, _) = body.kind;
+        if block.stmts.is_empty();
+        if let Some(inner_expr) = block.expr;
+        if let ExprKind::Match(
+            ref match_expr, ref match_arms, MatchSource::IfLetDesugar{ contains_else_clause }
+        ) = inner_expr.kind;
+        if !contains_else_clause;
+        // Ensure match_expr in `if let` statement is the same as the pat from the for-loop
+        if snippet(cx, match_expr.span, "..") == snippet(cx, pat.span, "..");
```
Comment (nahuakang) [inline clippy_lints/src/loops.rs:2009]: @camsteffen After failing some dogfood test, I realized that the `match_expr` and `pat` must be the same for this lint to trigger:
```rust
for pat in iter {
    if let Some(n) = match_expr {
        ... ...
    }
}
```
Otherwise the lint must handle even more complicated scenarios. However, I could only come up with a hacky approach to match the `snippet`s. Do you know how we could compare a `Pat` and an `Expr` to ensure they refer to the same thing?

Comment (camsteffen) [reply clippy_lints/src/loops.rs:2009]: First check that `pat` is a `PatKind::Binding` and get the `HirId`. Then check that `match_expr` is a `ExprKind::Path` that contains a `QPath::Resolved` with a `Res::Local` that contains the same `HirId` as the one in the `PatKind::Binding`.

File: clippy_lints/src/utils/higher.rs [hunk 4]
```diff
@@ -133,11 +134,11 @@ pub fn is_from_for_desugar(local: &hir::Local<'_>) -> bool {
...
 }
 
-/// Recover the essential nodes of a desugared for loop:
-/// `for pat in arg { body }` becomes `(pat, arg, body)`.
+/// Recover the entire span and the essential nodes of a desugared for loop:
+/// `for pat in arg { body }` becomes `(pat, arg, body)`. Return `(span, pat, arg, body)`.
 pub fn for_loop<'tcx>(
     expr: &'tcx hir::Expr<'tcx>,
-) -> Option<(&hir::Pat<'_>, &'tcx hir::Expr<'tcx>, &'tcx hir::Expr<'tcx>)> {
+) -> Option<(Span, &hir::Pat<'_>, &'tcx hir::Expr<'tcx>, &'tcx hir::Expr<'tcx>)> {
```
Comment (camsteffen) [inline clippy_lints/src/utils/higher.rs:141]: A nit but I think the Span is more commonly tacked on at the end of the tuple. Thanks for updating this!

File: clippy_lints/src/loops.rs (hunk 3 above)
Comment (nahuakang) [reply clippy_lints/src/loops.rs]: Thanks, this helps a lot! I saw the matching `HirId` but was not sure if that was the thing to match for!

File: clippy_lints/src/loops.rs [hunk 5]
```diff
@@ -494,8 +494,40 @@ declare_clippy_lint! {
     "there is no reason to have a single element loop"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for iteration of `Option`s with
+    /// a single `if let Some()` expression inside.
```
Comment (camsteffen) [inline clippy_lints/src/loops.rs:499]: Hmm this is a little too specific since it does not include `Result`.

> Checks for iteration where the items are "manually flattened" such as using `if let Some(inner) = item { .. }`.

File: clippy_lints/src/loops.rs [hunk 6]
```diff
@@ -494,8 +494,40 @@ declare_clippy_lint! {
...
+    /// for n in x {
+    ///     if let Some(n) = n {
+    ///         println!("{}", n);
+    ///     }
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let x = vec![Some(1), Some(2), Some(3)];
+    /// for n in x.iter().flatten() {
```
Comment (camsteffen) [inline clippy_lints/src/loops.rs:519]:
```suggestion
    /// for n in x.into_iter().flatten() {
```

File: clippy_lints/src/loops.rs [hunk 7]
```diff
@@ -1953,6 +1987,66 @@ fn check_for_single_element_loop<'tcx>(
...
+    body: &'tcx Expr<'_>,
+    span: Span,
+) {
+    if_chain! {
+        // Ensure the `if let` statement is the only expression in the for-loop
+        if let ExprKind::Block(ref block, _) = body.kind;
+        if block.stmts.is_empty();
+        if let Some(inner_expr) = block.expr;
+        if let ExprKind::Match(
+            ref match_expr, ref match_arms, MatchSource::IfLetDesugar{ contains_else_clause }
```
Comment (camsteffen) [inline clippy_lints/src/loops.rs:2005]: nit

```suggestion
            ref match_expr, ref match_arms, MatchSource::IfLetDesugar{ contains_else_clause: false }
```

**Review by camsteffen — CHANGES_REQUESTED** [review]
A couple minor things and then it looks good! Can you rebase onto master and squash into fewer commits?

File: clippy_lints/src/loops.rs (hunk 5 above)
Comment (nahuakang) [reply clippy_lints/src/loops.rs:499]: You're right! I completely forgot to edit this part.

**Comment by camsteffen** [comment]
@flip1995 ready for final review

File: clippy_lints/src/loops.rs [hunk 8]
```diff
@@ -1953,6 +1987,65 @@ fn check_for_single_element_loop<'tcx>(
...
+        if let Res::Local(match_expr_path_id) = match_expr_path.res;
+        if pat_hir_id == match_expr_path_id;
+        // Ensure the `if let` statement is for the `Some` variant of `Option` or the `Ok` variant of `Result`
+        if let PatKind::TupleStruct(QPath::Resolved(None, path), _, _) = match_arms[0].pat.kind;
+        if is_some_ctor(cx, path.res) || is_ok_ctor(cx, path.res);
+        let if_let_type = if is_some_ctor(cx, path.res) {
+            "Some"
+        } else {
+            "Ok"
+        };
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:2019]: Doing this check twice looks weird.

```suggestion
        let some_ctor = is_some_ctor(cx, path.res);
        let ok_ctor = is_ok_ctor(cx, path.res);
        if some_ctor || ok_ctor;
        let if_let_type = if some_ctor { "Some" } else { "Ok" };
```

File: clippy_lints/src/loops.rs [hunk 9]
```diff
@@ -1953,6 +1987,65 @@ fn check_for_single_element_loop<'tcx>(
...
+        // Determine if `arg` is `Iterator` or implicitly calls `into_iter`
+        let arg_ty = cx.typeck_results().expr_ty(arg);
+        if let Some(id) = get_trait_def_id(cx, &paths::ITERATOR);
+        if let is_iterator = implements_trait(cx, arg_ty, id, &[]);
+
+        then {
+            // Prepare the error message
+            let msg = format!("Unnecessary `if let` since only the `{}` variant of the iterator element is used.", if_let_type);
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:2027]:
```suggestion
            let msg = format!("unnecessary `if let` since only the `{}` variant of the iterator element is used", if_let_type);
```

@camsteffen @nahuakang error/help/... messages in Rust always start lower case and don't have punctuation at the end.

File: clippy_lints/src/loops.rs [hunk 10]
```diff
@@ -1953,6 +1987,65 @@ fn check_for_single_element_loop<'tcx>(
...
+
+            // Prepare the help message
+            let arg_snippet = snippet(cx, arg.span, "..");
+            let hint = if is_iterator {
+                format!("try: `{}.flatten()`", arg_snippet)
+            } else {
+                format!("try: `{}.into_iter().flatten()`", arg_snippet)
+            };
+
+            span_lint_and_help(
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:2037]: Why not use `span_lint_and_sugg` here. You already build the suggestion, why not suggest it?

File: tests/ui/manual_flatten.rs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    // Using the `None` variant should not trigger the lint
+    let z = vec![Some(1), Some(2), Some(3)];
+    for n in z {
+        if n.is_none() {
+            println!("Nada.");
+        }
+    }
+
+    // Using the `Err` variant should not trigger the lint
+    for n in y.clone() {
```
Comment (flip1995) [inline tests/ui/manual_flatten.rs:35]: A test for taking the vec by reference is missing AFAICS. So `for x in &y`. Then the suggestion should be to use `iter()` and not `into_iter()`

**Review by flip1995 — CHANGES_REQUESTED** [review]
@camsteffen great first review! You missed some things, that come up in some Clippy PRs, but those are things you'll pick up with experience :wink: 

@nahuakang Code LGTM so far. Mostly tests are missing (that may result in some changes that need to be made).

File: clippy_lints/src/loops.rs (hunk 9 above)
Comment (nahuakang) [reply clippy_lints/src/loops.rs:2027]: Duly noted for future commits!

File: clippy_lints/src/loops.rs (hunk 10 above)
Comment (camsteffen) [reply clippy_lints/src/loops.rs:2037]: Is it ok to suggest when an additional change is needed? (removing `if let`)

File: clippy_lints/src/loops.rs [hunk 12]
```diff
@@ -1996,52 +1996,82 @@ fn check_manual_flatten<'tcx>(
...
+                    arg.span,
+                    "..",
+                    &mut applicability,
+                );
+                // Determine if `arg` is by reference, an `Iterator`, or implicitly adjusted with `into_iter`
+                let hint = match arg.kind {
+                    ExprKind::AddrOf(_, _, arg_expr) => {
+                        format!("{}.iter().flatten()", snippet(cx, arg_expr.span, ".."))
+                    },
+                    ExprKind::MethodCall(_, _, _, _) | ExprKind::Path(QPath::Resolved(None, _)) => {
```
Comment (camsteffen) [inline clippy_lints/src/loops.rs:2048]: You can just check `expr_ty(arg).is_ref()` instead.

File: clippy_lints/src/loops.rs [hunk 13]
```diff
@@ -1996,52 +1996,82 @@ fn check_manual_flatten<'tcx>(
...
+                let arg_snippet = snippet_with_applicability(
+                    cx,
+                    arg.span,
+                    "..",
+                    &mut applicability,
+                );
+                // Determine if `arg` is by reference, an `Iterator`, or implicitly adjusted with `into_iter`
+                let hint = match arg.kind {
+                    ExprKind::AddrOf(_, _, arg_expr) => {
+                        format!("{}.iter().flatten()", snippet(cx, arg_expr.span, ".."))
```
Comment (camsteffen) [inline clippy_lints/src/loops.rs:2046]: So we know that `Vec::iter` is equivalent to `<&Vec as IntoIterator>::into_iter`, but unfortunately we cannot assume that this is true for any type. But, check it out, there's a `has_iter_method` function in utils just for that. If that returns `false`, I guess just bail cause I don't think you can make a suggestion without it looking ugly.

File: tests/ui/manual_flatten.stderr [hunk 14]
```diff
@@ -0,0 +1,44 @@
+error: unnecessary `if let` since only the `Some` variant of the iterator element is used
+  --> $DIR/manual_flatten.rs:6:5
+   |
+LL | /     for n in x {
+LL | |         if let Some(n) = n {
+LL | |             println!("{}", n);
+LL | |         }
+LL | |     }
+   | |_____^ help: try: `x.into_iter().flatten()`
```
Comment (flip1995) [inline tests/ui/manual_flatten.stderr:9]: So I forgot, that the `if let` also has to be removed. I would still use the `span_lint_and_sugg` with `MaybeIncorrect`, because it is a code suggestion, but use the span of `x` and not the complete for loop for the suggestion. This is possible with `span_lint_and_then`.

For future reference: Please add a test case where in `if let Some(x) = y` the unpacked binding name is different from the for loop binding. This will be of concern once someone tries to build a auto applicable suggestion.

File: clippy_lints/src/loops.rs [hunk 15]
```diff
@@ -1953,6 +1987,94 @@ fn check_for_single_element_loop<'tcx>(
...
+                } else if let Some(id) = get_trait_def_id(cx, &paths::ITERATOR) {
+                    let is_iterator = implements_trait(cx, arg_ty, id, &[]);
+                    if is_iterator {
+                        format!("{}.flatten()", arg_snippet)
+                    } else {
+                        format!("{}.into_iter().flatten()", arg_snippet)
+                    }
+                } else {
+                    return
+                };
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:2062]: I just discovered the function `make_iterator_snippet` in this module, which you should use here.
```suggestion
                let hint = make_iterator_snippet(cx, arg, &mut applicability);
```

File: clippy_lints/src/loops.rs [hunk 16]
```diff
@@ -2034,42 +2034,27 @@ fn check_manual_flatten<'tcx>(
...
+                        let sugg = format!("{}.flatten()", arg_snippet);
+                        diag.span_suggestion(
+                            arg.span,
+                            "try",
+                            sugg,
+                            Applicability::MaybeIncorrect,
+                        );
+                        diag.span_help(
+                            inner_expr.span,
+                            "also remove the `if let` statement in the for loop",
```
Comment (flip1995) [inline clippy_lints/src/loops.rs:2054]: NIT: I think it's clearer like this.

```suggestion
                            "...and remove the `if let` statement in the for loop",
```

File: tests/ui/manual_flatten.rs [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+        if let Ok(n) = n {
+            println!("{}", n);
+        }
+    }
+
+    // Test for loop over `Iterator` with `if let` expression
+    let z = vec![Some(1), Some(2), Some(3)];
+    let z = z.iter();
+    for n in z {
+        if let Some(m) = n {
```
Comment (flip1995) [inline tests/ui/manual_flatten.rs:41]:
```suggestion
        // Note: for an autofixable suggestion, the binding in the for loop has to take the 
        // name of the binding in the `if let`
        if let Some(m) = n {
```

**Review by flip1995 — COMMENTED** [review]
Two small things left.

(Remember to bless the tests again after those changes)

**Review by flip1995 — APPROVED** [review]

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Comment by nahuakang** [comment]
@flip1995 @camsteffen Thank you both so much for your patience and help!
